beps/00010-event-auditor/README.md

```diff
@@ -0,0 +1,221 @@
+
+## Design Details
+
+### Data Model for Audit Events
+
+To ensure consistency and facilitate regulatory compliance, the proposal suggests creating a shared package that defines a data model for audit events. This model consists of several key components.
+
+#### Actor Details Interface
+
+This interface defines the information related to the actor who triggered the logged event. It includes fields like actor ID, IP address, hostname, and user agent.
+
+```ts
+export type ActorDetails = {
+  actorId: string | null;
+  ip?: string;
+  hostname?: string;
+  userAgent?: string;
+};
+```
+
+#### Audit Request/Response Interfaces
+
+These interfaces define the structure of request and response data that might be included in the audit event. It's important to note that these interfaces exclude sensitive information like tokens from headers or other irrelevant details to avoid security risks.
+
+```ts
+export type AuditRequest = {
+  body: any;
+  url: string;
+  method: string;
+  params?: any;
+  query?: any;
+};
+
+export type AuditResponse = {
+  status: number;
+  body?: any;
+};
+```
+
+#### Audit Event Status Interfaces
+
+These interfaces define the possible statuses for an audit event entry. There are three options:
+
+```ts
+/**
+ * Indicates the event was successful.
+ */
+export type AuditEventSuccessStatus = { status: 'succeeded' };
+
+/**
+ * Indicates the event failed and includes details about the encountered errors.
+ */
+export type AuditEventFailureStatus = {
+  status: 'failed';
+  errors: ErrorLike[];
+};
+
+/**
+ * Indicates the event failed with errors that don't perfectly fit the expected structure.
+ */
+export type AuditEventUnknownFailureStatus = {
+  status: 'failed';
+  errors: unknown[];
+};
+
+export type AuditEventStatus = AuditEventSuccessStatus | AuditEventFailureStatus;
+```
+
+#### EventAuditor Interface
+
+This interface defines the functionalities of an `EventAuditor` class. This class provides methods for:
+- Extracting the actor ID from an Express request (if available).
+- Creating detailed audit event information based on provided options.
+- Logging an audit event with a specific level (info, debug, warn, or error).
+
+```ts
+/**
+ * Common fields of an audit event. Note: timestamp and pluginId are automatically added at event creation.
+ *
+ * @public
+ */
+export type AuditEventDetails = {
+  actor: ActorDetails;
+  eventName: string;
+  stage: string;
+  request?: AuditRequest;
+  response?: AuditResponse;
+  meta: JsonValue;
+  isAuditEvent: true;
+} & AuditEventStatus;
+
+export type AuditEventDetailsOptions = {
+  eventName: string;
+  stage: string;
+  metadata?: JsonValue;
+  response?: AuditResponse;
+  actorId?: string;
+  request?: Request;
+} & (AuditEventSuccessStatus | AuditEventUnknownFailureStatus);
+
+export type AuditEventOptions = {
+  eventName: string;
+  message: string;
+  stage: string;
```

Comment:
What's this? 😁

Reply:
Ah so the audit logging took lots of inspiration from the [kubernetes audit logging](https://kubernetes.io/docs/tasks/debug/debug-cluster/audit/) which keeps track of each stage of a request's execution.

> Each request can be recorded with an associated stage. The defined stages are:
> 
> - RequestReceived - The stage for events generated as soon as the audit handler receives the request, and before it is delegated down the handler chain.
> - ResponseStarted - Once the response headers are sent, but before the response body is sent. This stage is only generated for long-running requests (e.g. watch).
> - ResponseComplete - The response body has been completed and no more bytes will be sent.
> - Panic - Events generated when a panic occurred.

Reply:
Hmm, that seems very oriented towards just requests, do you think it makes sense to pivot towards that? Making sure that all important actions go through regular requests, and just make it an addition to the HTTP routing layer? With that we don't necessarily need any new service interface or plugin changes? It's gonna be a lot simpler and a more robust rollout, but it's not quite as capable at breaking down events. Still it might be plenty for what we need and a much simpler solution?

Reply:
Would it be possible to capture all these through requests?
Here's a quick list of areas we're looking to audit log for the core platform:
- Endpoints to determine what users are doing and who is accessing them (POST, GET, DELETE, etc.)
- Event Auditing Modifications made to resources in the database (
    - Entity deleted (important)
    - Entity added (important for users/groups)
    - Entity modifications (any suggestions on how this would be implemented?)
    - Tracking entity providers as well?
- Scaffolder Task Executions as they can modify resources in external services
    - Who triggered it
    - What actions/steps are being triggered (what is this task modifying/doing and what were they triggered with?)
    - Result of scaffolder task executions
- Auth
    - Who logged in
    - Did they login or logout or did their token expire, etc.
    - Logged into which auth provider, etc.
- Permission Framework:
    - Track authorization requests from users and tracks the authorization decisions for such requests

Reply:
@Rugvip any thoughts?

Reply:
Once I get a response I will update the BEP accordingly. 😄

Reply:
Hmm, I don't think we'll cover scaffolder task executions or auth with just requests. Could be that what we want to aim for is a bit of a hybrid then where most things are handled through a request middleware, but the rest is done via a common API, which is essentially what we got here atm.